lib/events/dynamoevents/dynamoevents.go

```diff
@@ -625,6 +781,233 @@ func (l *Log) getTableStatus(tableName string) (tableStatus, error) {
+// This means that the function may return an error without having processed
+// all data but no residual temporary or broken data is left and
+// the process can be resumed at any time by running this function again.
+//
+// Invariants:
+// - This function must be called after `createV2GSI` has completed successfully on the table.
+// - This function must not be called concurrently with itself.
+func (l *Log) migrateDateAttribute(ctx context.Context) error {
+	var total int64 = 0
+	var startKey map[string]*dynamodb.AttributeValue
+
+	for {
+		c := &dynamodb.ScanInput{
+			ExclusiveStartKey: startKey,
+			// Without consistent reads we may miss events as DynamoDB does not
+			// specify a sufficiently short synchronisation grace period we can rely on instead.
+			// This makes the scan operation slightly slower but the other alternative is scanning a second time
+			// for any missed events after an appropriate grace period which is far worse.
+			ConsistentRead: aws.Bool(true),
+			// 100 seems like a good batch size that compromises
+			// between memory usage and fetch frequency.
+			// The limiting factor in terms of speed is the update ratelimit and not this.
+			Limit:     aws.Int64(100),
+			TableName: aws.String(l.Tablename),
+			// Without the `date` attribute.
+			FilterExpression: aws.String("attribute_not_exists(CreatedAtDate)"),
+		}
+
+		// Resume the scan at the end of the previous one.
+		// This processes 100 events or 1 MiB of data at maximum
+		// which is why we need to run this multiple times on the dataset.
+		scanOut, err := l.svc.Scan(c)
+		if err != nil {
+			return trace.Wrap(convertError(err))
+		}
+
+		// For every item processed by this scan iteration we send an update action
+		// that adds the new date attribute.
+		for _, item := range scanOut.Items {
+			// Extract the UTC timestamp integer of the event.
+			timestampAttribute := item[keyCreatedAt]
+			var timestampRaw int64
+			err = dynamodbattribute.Unmarshal(timestampAttribute, &timestampRaw)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			// Convert the timestamp into a date string of format `yyyy-mm-dd`.
+			timestamp := time.Unix(timestampRaw, 0)
+			date := timestamp.Format(isoDateLayout)
+
+			attributes := map[string]interface{}{
+				// Value to set the date attribute to.
+				":date": date,
+			}
+
+			attributeMap, err := dynamodbattribute.MarshalMap(attributes)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			key := make(map[string]*dynamodb.AttributeValue)
+			key[keySessionID] = item[keySessionID]
+			key[keyEventIndex] = item[keyEventIndex]
+
+			c := &dynamodb.UpdateItemInput{
+				TableName:                 aws.String(l.Tablename),
+				Key:                       key,
+				ExpressionAttributeValues: attributeMap,
+				UpdateExpression:          aws.String("SET CreatedAtDate = :date"),
+			}
+
+			_, err = l.svc.UpdateItem(c)
+			if err != nil {
+				log.Infof("item fail data %q", item)
```

Comment:
nit: this looks like a leftover from debugging

Reply:
Removed